Library/Homebrew/cli_parser.rb

```diff
@@ -0,0 +1,280 @@
+module HomebrewCli
+  class Parser
+    attr_accessor :ignore_invalid_options
+    attr_reader :unnamed
+
+    def initialize(&block)
+      @specs = {}
+      @long = {}
+      @short = {}
+      @order = []
+      @constraints = []
+
+      instance_eval(&block)
+    end
+
+    def option(name, desc = "", opts = {})
+      opts[:desc] ||= desc
+
+      o = Option.create(name, desc, opts)
+
+      raise StandardError, "option #{o.name} already present" if @specs.member?(o.name) || @long[o.long]
+
+      @long[o.long] = o.name
+      @specs[o.name] = o
+
+      @order << [:opt, o.name]
+    end
+
+    def depends(*options)
+      options.each { |option| raise StandardError, "unknow #{option}" unless @specs[option] }
+      @constraints << [:depends, options]
+    end
+
+    def conflicts(*options)
+      options.each { |option| raise StandardError, "unknow #{option}" unless @specs[option] }
+      @constraints << [:conflicts, options]
+    end
+
+    def parse(cmdline = ARGV)
+      vals = {}
+      required = {}
+
+      @specs.each do |sym, opts|
+        required[sym] = true if opts.required?
+        vals[sym] = opts.default
+        vals[sym] = [] if opts.multi && !opts.default
+      end
+
+      given_args = {}
+      @unnamed = each_arg cmdline do |arg, params|
+        sym = case arg
+        when /^-([^-])$/ then @short[Regexp.last_match(1)]
+        when /^--([^-]\S*)$/ then @long[Regexp.last_match(1)]
+        else raise StandardError, "invalid syntax"
+        end
+        next nil if ignore_invalid_options && !sym
+        raise StandardError, "unknown argument #{arg}" unless sym
+
+        given_args[sym] ||= {}
+        given_args[sym][:arg] = arg
+        given_args[sym][:params] ||= []
+
+        num_params_taken = 0
+        unless params.empty?
+          if @specs[sym].single_arg?
+            given_args[sym][:params] << params[0, 1]
```

Comment:
Very confusing as to what this line is doing.

Reply:
Tried making it simpler now